### PR TITLE
Test/exp4p

### PR DESCRIPTION
**Pull request by taweihuang**
1. striatum/bandit/：完成bandit演算法。
(1) contextual bandit: LinUCB, Linear Thompson Sampling, EXP4.P
(2) multi-armed bandit: UCB1, EXP3

2. striatum/bandit/tests：完成所有bandit演算法的unittest。

3. simulation：模擬資料範例：對於所有演算法產生artificial dataset，並繪製parameter tunning curve。

4. 其他：新增setup.py。

### CODE REVIEW

**Comment by ianlini** [comment]
try to use English next time

**Comment by stegben** [comment]
Why not merge to master?

**Comment by taweihuang** [comment]
Ian, should I merge it to master?

**Comment by ianlini** [comment]
If you want to merge to master, you need to rebase on master and resolve the conflicts. 
It's really strange that you have conflicts here while you are the only developer. 
There must be something wrong. We can discuss it tomorrow.

**Comment by ianlini** [comment]
Seems that @yangarbiter add some strange code 17 days ago, please check it and resolve the conflicts, and then rebase on master.

**Comment by ianlini** [comment]
We should discuss your usage of Git tomorrow. I think you misunderstand something.

**Comment by ianlini** [comment]
BTW, if you want to merge to master, you need to create another PR, so I recommend you just merge to ucb1, and then merge ucb1 to master.

File: simulation/simulation_exp3.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+sys.path.append("..")
+from striatum.storage import history
+from striatum.storage import model
+from striatum.bandit import exp3
+
+
+class Exp3LinearPayoff:
+
+    def __init__(self, t, d, actions):
+        self.T = t
```
Comment (stegben) [inline simulation/simulation_exp3.py]: The uppercase t might not be a good choice for naming. Numpy transpose function is also uppercase t:
```python
import numpy as np

a = np.ones((10, 20))
b = a.T
```

File: simulation/simulation_linthompsamp.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+from striatum.bandit import linthompsamp
+
+
+class LinThompSampLinearPayoff:
+    def __init__(self, t, d, actions):
+        self.T = t
+        self.d = d
+        self.actions = actions
+
+    def data_simulation(self):
```
Comment (stegben) [inline simulation/simulation_linthompsamp.py]: There're several methods look almost the same among these simulation classes. I suggest to pull them out as single functions or build an abstract class.

Comment (stegben) [inline simulation/simulation_linthompsamp.py]: well, let it go. We'll do the refactor things in next PR

**Comment by taweihuang** [comment]
CPH had discussed my problem of git usage, and sorry for my misunderstanding~
I think the main reason for those conflicts is that I add "modelstorage" to store parameters for all bandit algorithms.

File: simulation/simulation_exp3.py (hunk 1 above)
Comment (taweihuang) [inline simulation/simulation_exp3.py]: okay

File: striatum/bandit/linucb.py [hunk 3]
```diff
@@ -1,99 +1,157 @@
...
-
-        history_id = self.storage.add_history(None, action, reward=None)
-        self.last_history_id = history_id
-
-        return history_id, action
+            self.linucb_ = self.linucb
+            self.linucb_.next()
+            action_max = self.linucb_.send(context)
+        else:
+            self.linucb_.next()
```
Comment (stegben) [inline striatum/bandit/linucb.py]: It's python2 style generator, which will raise error in python 3+. Try ```six``` instead:
```python
import six

...
six.next(self.linucb_)
```

Document: https://pythonhosted.org/six/#six.next
StackOverflow: http://stackoverflow.com/questions/1073396/is-generator-next-visible-in-python-3-0

File: striatum/bandit/linucb.py [hunk 4]
```diff
@@ -1,99 +1,157 @@
...
+        theta = self._modelstorage.get_model()['theta']
+
+        for key in actions:
+            if key not in self._actions:
+                matrix_a[key] = np.identity(self.d)
+                matrix_ainv[key] = np.identity(self.d)
+                b[key] = np.zeros((self.d, 1))
+                theta[key] = np.zeros((self.d, 1))
+
+        self._actions = actions
```
Comment (stegben) [inline striatum/bandit/linucb.py]: Wouldn't it replace the original actions?
Maybe use ```append```?

Comment (ianlini) [inline striatum/bandit/linucb.py]: This is definitely wrong because there is no `self._actions`

**Comment by stegben** [comment]
The rest LGTM. Some refactor issue can be solved in the future.
@taweihuang do you need help rebasing?

File: striatum/bandit/linucb.py (hunk 4 above)
Comment (ianlini) [inline striatum/bandit/linucb.py]: ah...sorry, it's in the abstract class

**Comment by ianlini** [comment]
@stegben Yes, please help him and check whether the conflicts are reasonable.

**Comment by taweihuang** [comment]
@stegben  Definitely need!!!! Please help me at your convenience!

File: striatum/bandit/linucb.py (hunk 3 above)
Comment (taweihuang) [inline striatum/bandit/linucb.py]: Get it~ I currently use python 2 so I haven't find it, thx!

File: striatum/bandit/linucb.py (hunk 4 above)
Comment (taweihuang) [inline striatum/bandit/linucb.py]: I use `extend` instead to allow multiple new actions.

**Pull request merged**